STIPS/STIP-005.md

```diff
@@ -0,0 +1,1983 @@
++ usdcToWithdraw = externalPositionUnit * _setTokenQuantity
++ setToken.invokeWithdraw(usdcToWithdraw)
+-----
+
+> **deposit**: 
+
+```solidity
+function _deposit(
+  ISetToken _setToken, 
+  uint256 _collateralQuantityUnits
+) 
+  external
+  nonReentrant
+  onlyManagerAndValidSet(_setToken)
+```
+
++ notionalQuantity = _collateralQuantityUnits * _setToken.totalSupply
++ token = collateralToken[_setToken]
++ decimals = token.decimals()
++ notionalQuantityInCollateralDecimals = _formatCollateralToken(notionalQuantity, decimals)
++ _setToken.invokeApprove(collateralToken, protocolAddresses.vault, notionalQuantityInCollateralDecimals)
++ _setToken.invokeDeposit(protocolAddresses.vault, token, notionalQuantityInCollateralDecimals)
+----
+
+> **withdraw**: 
+
+```solidity
+function _withdraw(
+  ISetToken _setToken, 
+  uint256 _collateralQuantityUnits
+) 
+  external
+  nonReentrant
+  onlyManagerAndValidSet(_setToken)
+```
++ notionalQuantity = _collateralQuantityUnits * _setToken.totalSupply
++ token = collateralToken[_setToken]
++ decimals = token.decimals()
++ notionalQuantityInCollateralDecimals = _formatCollateralToken(notionalQuantity, decimals)
++ _setToken.invokeWithdraw(protocolAddresses.vault, token, notionalQuantityInCollateralDecimals)
+-----
+
+> **trade**: 
+
+```solidity
+function trade(
+  ISetToken _setToken,
+  address _debitToken,  
+  int256 _debitQuantityUnits, 	  // (negative)
+  address _creditToken,
+  int256 _minCreditQuantityUnits, // (positive)
+  bytes memory _data        
+)
+  external
+  nonReentrant
+  onlyManagerAndValidSet(_setToken)
+  returns (uint256 deltaBase, uint256 deltaQuote)
+```
+
+Format Trade parameters
+
++ tradeParams = 
+  + collateral = collateralToken[_setToken]
+  + isBaseToQuote = _debitToken != collateral
+  + baseToken = (isBaseToQuote) ? _creditToken : _debitToken 
+  + amount = abs(_debitQuantityUnits * _setToken.totalSupply)
+  + oppositeAmountBound = abs(_creditQuantityUnits * _setToken.totalSupply)
+  + isExactInput = true
+  + (
+      deadline,
+      sqrtPriceLimitX96,
+      referralCode
+    ) = abi.decode(bytes, (bool,uint256,uint160,bytes32))
+ 
+Execute trade and accrue fees. *minCreditQuantityUnits* check is performed by PerpV2 
++ (deltaBase, deltaQuote) = setToken.invokeOpenPosition(_setToken, protocolAddresses.clearingHouse, tradeParams)
++ _accrueProtocolFee(_setToken, deltaQuote)
+
+Check position balance and update position data structure with additions or removals
++ positions = getPositions(_setToken)
++ baseBalance = protocolAddresses.AccountBalance.getPositionSize(_setToken, tradeParams.baseToken) 
++ if (baseBalance == 0) 
+  + removePosition(_setToken, tradeParams.baseToken)
++ elseif (baseBalance > 0 && !positions.contains(tradeParams.baseToken)) 
+  + addPosition(_setToken, tradeParams.baseToken)
+
+Return deltas
++ return (deltaBase, deltaQuote)
+------
+
+> **lever**: 
+
+```solidity
+function lever(
+  ISetToken _setToken,
+  uint256 _quoteQuantityUnits,
```

Comment:
It'd still be nice to have the token addresses in the interface here. From a functionality perspective I feel like we may want the ability to isolate certain assets that we want to trade and not just apply the same changes across all assets.

Reply:
I think we want both `lever` and `delever` to have as close to the same interfaces as we can to ALM and CLM:
```
    function lever(
        ISetToken _setToken,
        IERC20 _borrowAsset,
        IERC20 _collateralAsset,
        uint256 _borrowQuantityUnits,
        uint256 _minReceiveQuantityUnits,
    )
```
This would make the most sense to me, a little unclear how we seem to be spending some amount of quote quantity but then checking it against some `receiveQuoteQuantity`.